### PR TITLE
Sentry Refactor and Bug Fix

### PR DESCRIPTION
**Pull request by mueschm**
## Overview
This diff cleans up some of the Sentry code as well as fixes how it handles exit codes.

## Changes
1. Migrated most code into the sentry class and out of base command
2. Added tests to make sure exit codes get set on error for regsiter
3. Cleaned up some of the logic to decouple Oclif and Sentry
4. Switched the debug logs to now use Oclif debug. We will most likely migrate off those, but this is not a blocker. To view the logs just put `DEBUG=architect:* ` before any command.

Worked on it in conjunction with Devin.

### Basic Tests
`npm run test`
### Error Code
`architect-dev register .
/examples/react-app/architect.yml --architecture not-real`
This will fail since the architecture is not real. Then check the error code with
`echo $?` which should return `1`.
You can also see the error logs appear in Sentry.
### Working
`architect-dev register .
/examples/react-app/architect.yml --architecture not-real`
`echo $?` which should return `0`.

***Edit:** Gitlab issue: https://gitlab.com/architect-io/architect-cli/-/issues/472*

### CODE REVIEW

File: src/base-command.ts [hunk 1]
```diff
@@ -112,84 +111,17 @@ export default abstract class BaseCommand extends Command {
...
-      this.warn('Failed to get command metadata');
-    }
-
-    await this.sentry.endSentryTransaction(!(await this.disable_sentry_recording()), err);
-
-    if (!err) {
-      return await super.finally(err);
-    }
-    return await super.catch(err);
+    await this.sentry?.endSentryTransaction(!(await this.disable_sentry_recording()), await this.parse(calling_class), calling_class, err);
```
Comment (tjhiggins) [inline src/base-command.ts:116]: wrap in try catch

File: src/base-command.ts [hunk 2]
```diff
@@ -112,84 +111,17 @@ export default abstract class BaseCommand extends Command {
...
+    // Oclif supers go as the return
+    return super.finally(err);
   }
 
+  async catch(err: any): Promise<void> {
+    if (err.oclif && err.oclif.exit === 0) return;
+    await this.sentry?.catch(err);
```
Comment (tjhiggins) [inline src/base-command.ts:123]: wrap in try catch

**Review by tjhiggins — CHANGES_REQUESTED** [review]

File: src/base-command.ts [hunk 3]
```diff
@@ -37,6 +33,19 @@ export default abstract class BaseCommand extends Command {
...
+  private async createSentry() {
+    try {
+      this.sentry = await SentryService.create(this.app, this.constructor as any, this.debug.bind(this));
+    } catch (e) {
+      this.debug('SENTRY: an error occurred creating a new instance of SentryService');
+    }
+  }
+
+  private async loginRequired() {
+    await this.createSentry();
```
[the remark this answers is not in the document]
Comment (mueschm) [reply src/base-command.ts:51]: We can but it gets a bit weird with the transaction system. Normally we want to associate the user with the event, but if they are not logged in yet we dont have that. So two choices.
1. Do what we are doing now.
2. Create the transaction and then once logged in attach the metadata.

This code actually looked cleaner to me so I wound up going with this. The only reason we would even need this is if login errors occur. So its kind of a special case anyway compared to the rest of the system.

[the remark this answers is not in the document]
Comment (tjhiggins) [reply src/base-command.ts:51]: Hmm I guess I was worried about exceptions we wouldn't catch prior to login.

[the remark this answers is not in the document]
Comment (mueschm) [reply src/base-command.ts]: Alright, it now happens on the constructor. Any error that happens from that point will be captured.

**Pull request merged**